### PR TITLE
Split forbidden host/domain code points, and add all C0 controls and add U+007F to the latter

### PR DESCRIPTION
**Pull request by karwa**
Attempt at resolving #627 and #319

- [x] At least two implementers are interested (and none opposed):
   * Chrome (Presumably - See linked issue)
   * WebKit (Presumably - See linked issue)
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/32524
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1291564
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1752320
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=235709
   * Deno: https://github.com/denoland/deno/issues/13509
   * Node.js: https://github.com/nodejs/node/issues/41717

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

***
<a href="https://whatpr.org/url/685.html" title="Last updated on Jan 28, 2022, 8:59 AM UTC (3229929)">Preview</a> | <a href="https://whatpr.org/url/685/68dc61c...3229929.html" title="Last updated on Jan 28, 2022, 8:59 AM UTC (3229929)">Diff</a>

### CODE REVIEW

**Review by annevk — APPROVED** [review]
This looks good to me. I don't think the conformance requirements end up needing adjusting (and it seems like this fixes a small problem for "valid opaque-host string"), although for "valid domain string" it's a bit unclear due to the imperative nature of its definition.

I'd appreciate a second review.

**Comment by domenic** [comment]
> Chrome (Presumably - See linked issue)

Which part of which linked issue are you referring to?

**Comment by domenic** [comment]
Hmm. But https://wpt.fyi/results/url?diff&filter=ADC&run_id=5656521351364608&run_id=5648068402741248 shows many new failing tests in Chrome. Are all of those tests that would already fail with the current spec? Maybe it'd be best to separate out the tests PR into one that tests the current spec, and one that tests the change here, so that we can actually see if the change here causes any deltas from the current spec?

**Comment by karwa** [comment]
There are more failures because there are more tests, but for the part which is changed by this spec change - domains containing (raw or encoded) code points in the range `\u0000-\u0001F` or `\u007F` - Chrome passes them all.

If I changed the tests to the current spec, there would be even more failures for _all browsers_. None of them implement the current spec. WebKit was the last one, but @achristensen07 recently changed made the change to reject those code-points (for all hosts, not just domains, which is why WebKit still has one failure, but the consensus from the issue seemed to be to leave those alone for now):

> I think there is a consensus that the forbidden host code points should include all C0 controls and U+007F, and that now matches the behavior of all browsers. Let's make that change to the specification and then make a separate issue about what we want to do with opaque hosts, where there is less consensus and less spec compliance.

https://github.com/whatwg/url/issues/627#issuecomment-981906774

**Comment by domenic** [comment]
> If I changed the tests to the current spec, there would be even more failures for all browsers

Right. That would be helpful. Then we could double-check that the new tests, specific to the changes in this PR, are all passing in Chromium.

**Comment by karwa** [comment]
OK, I've put up https://github.com/web-platform-tests/wpt/pull/32539

Same tests as linked for this change, but with expected values that match the current spec.

**Comment by domenic** [comment]
Editorially LGTM, let's just rebase https://github.com/web-platform-tests/wpt/pull/32524 on WPT master, confirm that the result is more in alignment with browsers, and then we should be good to merge. Thanks @karwa for doing the work, including the testing against whatwg-url.

**Comment by annevk** [comment]
@karwa do you want to file the implementation bugs? https://github.com/whatwg/meta/blob/master/MAINTAINERS.md has pointers.

**Comment by karwa** [comment]
@annevk bugs have been filed.

**Comment by annevk** [comment]
Thanks a lot @karwa for your work on tackling this as well as providing all those great tests!

**Pull request merged**
